src/main/java/de/tum/cit/aet/artemis/programming/service/UserSshPublicKeyService.java

```diff
@@ -0,0 +1,141 @@
+
+    private final UserSshPublicKeyRepository userSshPublicKeyRepository;
+
+    public UserSshPublicKeyService(UserSshPublicKeyRepository userSshPublicKeyRepository) {
+        this.userSshPublicKeyRepository = userSshPublicKeyRepository;
+    }
+
+    /**
+     * Creates a new SSH public key for the specified user, ensuring that the key is unique
+     * based on its SHA-512 hash fingerprint. If the key already exists, an exception is thrown.
+     *
+     * @param user         the {@link User} for whom the SSH key is being created.
+     * @param keyEntry     the {@link AuthorizedKeyEntry} containing the SSH public key details, used to resolve the {@link PublicKey}.
+     * @param sshPublicKey the {@link UserSshPublicKey} object containing metadata about the SSH key such as the key itself, label, and expiry date.
+     */
+    public void createSshKeyForUser(User user, AuthorizedKeyEntry keyEntry, UserSshPublicKeyDTO sshPublicKey) throws GeneralSecurityException, IOException {
+        PublicKey publicKey = keyEntry.resolvePublicKey(null, null, null);
+        String keyHash = HashUtils.getSha512Fingerprint(publicKey);
+
+        if (userSshPublicKeyRepository.findByKeyHash(keyHash).isPresent()) {
+            throw new BadRequestAlertException("Key already exists", "SSH key", "keyAlreadyExists", true);
+        }
+
+        UserSshPublicKey newUserSshPublicKey = new UserSshPublicKey();
+        newUserSshPublicKey.setUserId(user.getId());
+        newUserSshPublicKey.setPublicKey(sshPublicKey.publicKey());
+        newUserSshPublicKey.setKeyHash(keyHash);
+        setLabelForKey(newUserSshPublicKey, sshPublicKey.label());
+        newUserSshPublicKey.setCreationDate(ZonedDateTime.now());
+        newUserSshPublicKey.setExpiryDate(sshPublicKey.expiryDate());
+        userSshPublicKeyRepository.save(newUserSshPublicKey);
+    }
+
+    /**
+     * Sets the label for the provided SSH public key. If the given label is null or empty,
+     * the label is extracted from the public key or defaults to a predefined value.
+     *
+     * @param newSshPublicKey the {@link UserSshPublicKey} for which the label is being set.
+     * @param label           the label to assign to the SSH key, or null/empty to use the default logic.
+     * @throws BadRequestAlertException if the key label is longer than 50 characters
+     */
+    private void setLabelForKey(UserSshPublicKey newSshPublicKey, String label) {
+        if (StringUtils.isBlank(label)) {
+            String[] parts = newSshPublicKey.getPublicKey().split("\\s+");
+
+            // we are only interested in the comment of the key. A typical key looks like this, the key prefix, the actual key and then the comment:
+            // ssh-rsa AAAAB3NzaC1yc2EAAAADAYVTLQ== comment
+            if (parts.length >= 3) {
+                label = String.join(" ", Arrays.copyOfRange(parts, 2, parts.length));
+            }
+            else {
+                label = KEY_DEFAULT_LABEL;
```

Comment:
What happens when the user already has a key named `Key 1`? I'd love if you could actually count the amount of keys and use the correct number then 😍
```suggestion
                label = "Key " + userSshPublicKeyRepository.findAllByUserId(user.getId()).stream().map(UserSshPublicKeyDTO::of).toList().length + 1;
```

Reply:
aah for some reason the case where a user doesn't set a label and does not have a comment in the key, didn't really come to my mind. But yeah, for that case this would make sense, as otherwise all keys are just called "Key 1". I'll add that :+1: